### ISSUE TITLE
devp2p: node discovery ping echo hash bug

### ISSUE THREAD

**Issue by holgerd77** [paritytech/parity#8038]
Attributes to: current ``master``, os-and sync-independent

### Intro

Hi, this is Holger from the ``ethereumjs`` team. We had an [issue](https://github.com/ethereumjs/ethereumjs-devp2p/issues/17) on our devp2p implementation (``ethereumjs-devp2p``) not being able to connect to Parity clients, which in the end turned out to be a bug on the Parity side (I now fixed with a hotfix for the moment).

### Bug Description

When sending a ``DPT`` ``ping`` message to Parity the echoed hash is not composed correctly. In the ``on_ping()`` method in ``parity/util/network/src/discovery.rs`` only the RLP-enocded message content [is taken](https://github.com/paritytech/parity/blob/e95b09348386d01b71901365785c5fa3aa2f7a6d/util/network/src/discovery.rs#L421) for creating the hash:

```rust
fn on_ping(&mut self, rlp: &UntrustedRlp, node: &NodeId, from: &SocketAddr) -> Result<Option<TableUpdates>, Error> {
   // ...
   let hash = keccak(rlp.as_raw());
   let mut response = RlpStream::new_list(2);
   dest.to_rlp_list(&mut response);
   response.append(&hash);
   self.send_packet(PACKET_PONG, from, &response.drain());
}
```

This differs from the specification (see ["Node Discovery"](https://github.com/ethereum/devp2p/blob/master/rlpx.md#node-discovery)) where the echo hash is created by also hashing ``signature`` and ``packet-type``:

```
hash: sha3(signature || packet-type || packet-data)

Pong packet-type: 0x02
struct Pong
{
	Endpoint to;
	h256 echo;
	uint32_t timestamp;
};
```

### Consequences/Severity

The bug is **potentially severe**, totally preventing connections (at least uni-directional) to other client implementations, depending if the hash is checked/used on the specific client implementation side.

### Possible Solutions

Parity is currently [not validating](https://github.com/paritytech/parity/blob/e95b09348386d01b71901365785c5fa3aa2f7a6d/util/network/src/discovery.rs#L432) the echoed hash in ``on_pong()`` method. So it should be possible to update the echo hash to the correct specification without introducing connectivity problems with older-version Parity clients. This is not tested though.

**Comment by andresilva** [comment]
@holgerd77 Can you check that this is fixed with the latest master?

**Comment by holgerd77** [comment]
@andresilva Phew, I am neither Parity nor Rust guy and would do the whole ``master`` deployment setup for the first time. On the other hand would be an interesting task/Parity get-to-know challenge, so I will try to get this in within the week. Can't promise though, have other things on the table.

**Comment by andresilva** [comment]
Sorry, totally understand the commitment that would be necessary. We publish nightly binaries so you can grab one of those instead (https://vanity-service.parity.io/parity-binaries?format=markdown&version=nightly), a new one should be available by tomorrow that includes the fix. 👍

**Comment by holgerd77** [comment]
Hi, I will be able to look into this at the end of the week. Hope this still helps.

**Comment by holgerd77** [comment]
Hi @andresilva, ok, found some time today to check this, here are the results:

---

### Parity v1.8.11, local node, no internet connection

#### ethereumjs-devp2p Parity Hotfix (taking the wrong hash) enabled:
Working! (expected)

```
devp2p:dpt attempt adding peer 127.0.0.1:30303 +0ms
devp2p:dpt:server send ping to 127.0.0.1:30303 (peerId: undefined) +1ms
devp2p:dpt:server received pong from 127.0.0.1:30303 (peerId: bba012ada9a93852e3b7b351fa2768b7484a75b0b5a6429bb7ac3a6fb916a083afed4c2eacdabe645925473d6526dafac8745b2746cb05e73c4c2871685d8f07) +178ms
devp2p:rlpx connect to 127.0.0.1:30303 (id: bba012ada9a93852e3b7b351fa2768b7484a75b0b5a6429bb7ac3a6fb916a083afed4c2eacdabe645925473d6526dafac8745b2746cb05e73c4c2871685d8f07) +3ms
Add peer: 127.0.0.1:30303 Parity/v1.8.11-stable-21522ff-20180227/x86_64-macos/rustc1.24.0 (eth63) (total: 1)
```

#### Parity Hotfix disabled
Not working! (expected)

```
devp2p:dpt attempt adding peer 127.0.0.1:30303 +0ms
devp2p:dpt:server send ping to 127.0.0.1:30303 (peerId: undefined) +2ms
devp2p:dpt:server received pong from 127.0.0.1:30303 (peerId: bba012ada9a93852e3b7b351fa2768b7484a75b0b5a6429bb7ac3a6fb916a083afed4c2eacdabe645925473d6526dafac8745b2746cb05e73c4c2871685d8f07) +238ms
devp2p:rlpx refill connections.. queue size: 0, open slots: 25 +10s
devp2p:dpt:server ping timeout: 127.0.0.1:30303 undefined +19ms
error on connection to local node: Error: Timeout error: ping 127.0.0.1:30303
    at Timeout._onTimeout (/Users/hdrewes/Documents/DEV/EthereumJS/ethereumjs-devp2p/src/dpt/server.js:77:27)
    at ontimeout (timers.js:386:14)
    at tryOnTimeout (timers.js:250:5)
    at Timer.listOnTimeout (timers.js:214:5)
```

### Parity v1.11.0 (nightly binary, 2018-03-22), local node, no internet connection

Binary chosen: darwin | x86_64 | [parity_1.11.0_macos_macos.pkg](http://d1h4xl4cr1h0mo.cloudfront.net/nightly/x86_64-apple-darwin/parity_1.11.0_macos_macos.pkg) | 252ec483d0837b98a1dcdad816cb591f

#### ethereumjs-devp2p Parity Hotfix (taking the wrong hash) enabled:
Working! (expected)

```
devp2p:dpt attempt adding peer 127.0.0.1:30303 +0ms
devp2p:dpt:server send ping to 127.0.0.1:30303 (peerId: undefined) +2ms
devp2p:dpt:server received pong from 127.0.0.1:30303 (peerId: bba012ada9a93852e3b7b351fa2768b7484a75b0b5a6429bb7ac3a6fb916a083afed4c2eacdabe645925473d6526dafac8745b2746cb05e73c4c2871685d8f07) +306ms
devp2p:rlpx connect to 127.0.0.1:30303 (id: bba012ada9a93852e3b7b351fa2768b7484a75b0b5a6429bb7ac3a6fb916a083afed4c2eacdabe645925473d6526dafac8745b2746cb05e73c4c2871685d8f07) +3ms
devp2p:rlpx connected to 127.0.0.1:30303, handshake waiting.. +43ms
Add peer: 127.0.0.1:30303 Parity/v1.11.0-nightly-a60d0e4-20180321/x86_64-macos/rustc1.24.1 (eth63) (total: 1)
```

#### Parity Hotfix disabled
Still working! :-) (expected)

```
devp2p:dpt attempt adding peer 127.0.0.1:30303 +0ms
devp2p:dpt:server send ping to 127.0.0.1:30303 (peerId: undefined) +1ms
devp2p:dpt:server received pong from 127.0.0.1:30303 (peerId: bba012ada9a93852e3b7b351fa2768b7484a75b0b5a6429bb7ac3a6fb916a083afed4c2eacdabe645925473d6526dafac8745b2746cb05e73c4c2871685d8f07) +238ms
devp2p:rlpx connect to 127.0.0.1:30303 (id: bba012ada9a93852e3b7b351fa2768b7484a75b0b5a6429bb7ac3a6fb916a083afed4c2eacdabe645925473d6526dafac8745b2746cb05e73c4c2871685d8f07) +2ms
devp2p:rlpx connected to 127.0.0.1:30303, handshake waiting.. +38ms
Add peer: 127.0.0.1:30303 Parity/v1.11.0-nightly-a60d0e4-20180321/x86_64-macos/rustc1.24.1 (eth63) (total: 1)
```

For the last check I double checked that I really disabled the hotfix (by adding an extra console output after commenting out) so to make sure that I didn't forgot to save or something.

So I can confirm that the fix is working! 😄

**Comment by andresilva** [comment]
Great! Thanks a lot for the initial report and for taking the extra time to verify that the fix is working properly. We'll include this in the next release. 🙌

**Comment by gsalgado** [comment]
Since this is a trivial bugfix, would it make sense to backport it to 1.10 as well?

**Comment by 5chdn** [comment]
this made it into 1.10

**Comment by gsalgado** [comment]
Are you sure, @5chdn ? I don't see it on https://github.com/paritytech/parity/blob/v1.10.9/util/network/src/discovery.rs#L382 but maybe that's not where 1.10 was cut from?

**Comment by gsalgado** [comment]
Also, I've tested locally against a 1.10.8 instance and it is not fixed there either

**Comment by 5chdn** [comment]
We are not releasing any further 1.10 version anyways. Could you check 1.11?

**Comment by gsalgado** [comment]
It is fixed in 1.11, the problem is that there are lots of 1.10 nodes out there and py-evm cannot use them for node discovery because of this bug.
